pkg/datasource.go

```diff
@@ -153,29 +161,66 @@ func (o *OCIDatasource) QueryData(ctx context.Context, req *backend.QueryDataReq
-	// 	return &backend.QueryDataResponse{}, err
-	// }
-
-	// //o.logger.Error("ts.Com is " + ts.Compartment)
-	// listMetrics := monitoring.ListMetricsRequest{
-	// 	CompartmentId: common.String(ts.Compartment),
-	// }
-	// reg := common.StringToRegion(ts.Region)
-	// o.metricsClient.SetRegion(string(reg))
-	// res, err := o.metricsClient.ListMetrics(ctx, listMetrics)
-	// if err != nil {
-	// 	return &backend.QueryDataResponse{}, err
-	// }
-	// status := res.RawResponse.StatusCode
-	// if status >= 200 && status < 300 {
-	// 	return &backend.QueryDataResponse{}, nil
-	// }
-	// return nil, errors.Wrap(err, fmt.Sprintf("list metrircs failed %s %d", spew.Sdump(res), status))
+
+	// GrafanaCommonRequest - captures the common parts of the search and metricsRequests
+	// the structure is used here to retrieve the TenancyOCID and CompartmentOCID
+	var ts GrafanaCommonRequest
+	query0 := req.Queries[0]
+
+	if err := json.Unmarshal(query0.JSON, &ts); err != nil {
+		return &backend.QueryDataResponse{}, err
+	}
+	o.logger.Debug("Testing OCI logs datasource", "TenancyOCID", ts.TenancyOCID)
+	o.logger.Debug("Testing OCI logs datasource", "CompartmentOCID", ts.Compartment)
+
+	listLogsGroup := logging.ListLogGroupsRequest{
+		CompartmentId: common.String(ts.Compartment),
+	}
+	listLogsGroups, err := o.loggingMgmtClient.ListLogGroups(ctx, listLogsGroup)
+	if err == nil {
+		for _, loggroupitem := range listLogsGroups.Items {
+			if &loggroupitem.Id != nil {
+				listLog := logging.ListLogsRequest{
+					LogGroupId: common.String(*loggroupitem.Id),
+				}
+				listLogs, err := o.loggingMgmtClient.ListLogs(ctx, listLog)
+				if err == nil {
+					for _, logitem := range listLogs.Items {
+						if &logitem.Id != nil {
+							query := `search "` + ts.Compartment + `/` + *loggroupitem.Id + `/` + *logitem.Id + `"`
+							t := time.Now()
+							t2 := t.Add(-time.Minute * 30)
+							start, _ := time.Parse(time.RFC3339, t2.Format(time.RFC3339))
+							end, _ := time.Parse(time.RFC3339, t.Format(time.RFC3339))
+							request := loggingsearch.SearchLogsRequest{SearchLogsDetails: loggingsearch.SearchLogsDetails{SearchQuery: common.String(query),
+								TimeStart:         &common.SDKTime{Time: start},
+								TimeEnd:           &common.SDKTime{Time: end},
+								IsReturnFieldInfo: common.Bool(false)},
+								Limit: common.Int(10)}
+							res, err := o.loggingSearchClient.SearchLogs(ctx, request)
+							if err == nil {
+								status := res.RawResponse.StatusCode
+								o.logger.Debug("status :", "status", status)
+								if status >= 200 && status < 300 {
+									return &backend.QueryDataResponse{}, nil
+								}
+							} else {
+								// return &backend.QueryDataResponse{}, err
+								continue
+							}
+						}
+					}
+				} else {
+					// return &backend.QueryDataResponse{}, err
+					continue
+				}
+			}
+		}
+	} else {
+		return &backend.QueryDataResponse{}, err
+	}
+
+	return &backend.QueryDataResponse{}, err
```

Comment:
There are a number of paths through this code where the `err` variable will not be set to a reasonable value (or at all). Specifically the cases I see where there is no proper error being defined are when the results of the `ListLogGroups()` and `ListLogs()` operations are processed and the returned list is empty.